src/redprl/refiner_kit.fun

```diff
@@ -31,9 +31,32 @@ struct
       end
   end
 
+  fun abstractEvidence (I : (sym * psort) list, H) m =
+    let
+      val (us, sigmas) = ListPair.unzip I
+      val (xs, taus) = Hyps.foldr (fn (x, jdg, (xs, taus)) => (x::xs, CJ.synthesis jdg::taus)) ([],[]) H
+    in
+      Abt.checkb (Abt.\ ((us, xs), m), ((sigmas, taus), Abt.sort m))
+    end
+
+  fun #> (psi, (I, H, m)) =
+    Lcf.|> (psi, abstractEvidence (I, H) m)
+  infix #>
+
+  val trivial = Syn.into Syn.AX
+
+  fun orelse_ (t1, t2) alpha = Lcf.orelse_ (t1 alpha, t2 alpha)
+  infix orelse_
+
+  fun >:+ (tel, (l : (label * 'a) list)) : 'a telescope =
+    List.foldl (fn (g, t) => t >: g) tel l
+  infix 5 >:+
 
-  fun appendListOfGoals (tel, (l : (label * 'a) list)) : 'a telescope =
-    List.foldl (fn (g, l) => l >: g) tel l
+  (* this is a hack till we have a nice way to pre-compose Equality.Symmetry *)
+  fun catJdgFlip (CJ.EQ_TYPE (a, b)) = CJ.EQ_TYPE (b, a)
+    | catJdgFlip (CJ.EQ ((a, b), ty)) = CJ.EQ ((b, a), ty)
+  fun catJdgFlipWrapper tactic alpha (IH >> jdg) =
+    tactic alpha (IH >> catJdgFlip jdg)
```

Comment:
After getting the pretty printer set up, I can make my next task to make it possible to write real composed tactics in the RedPRL source. This is not a technically difficult thing to do—it just means rearranging some modules.